atom/github
# Move "Undo Discard" button behind context menu

### Description of the Change


This PR removes the `Undo Discard` button and hides this functionality behind a context menu that is accessed from the Unstaged Changes list **header**. 

<img width="308" alt="git_ ___src_test-repo" src="https://user-images.githubusercontent.com/7910250/45997160-8b7f6280-c053-11e8-928a-437eb76478b8.png">
<img width="397" alt="fullscreen_9_24_18__11_43_pm" src="https://user-images.githubusercontent.com/7910250/45997207-a9e55e00-c053-11e8-915a-68a0ca4c59dd.png">

Note: A similar context menu currently exists in the Merge Conflicts list header for bulk resolve actions.

When multiple files are selected, the context menu item is pluralized to `Discard Changes in Selected Files`. I chose this more verbose description (rather than simply `Discard Changes`) to make it clear that this would discard changes for the Unstaged Changes file list, and not for the file diff view which has separate discard state.

The Stage/Unstage all buttons no longer disappear when there are no changes. Now they simply become disabled. This gives something for the new kebab menu to sit up against.

### Alternate Designs


Initially, I moved the `Undo Discard` button into the header and hid the text when the width was too narrow (see https://github.com/atom/github/commit/b82e732ca7e1d6e642fdb4f133447064647d51a5):

<img width="370" alt="git_ ___src_test-repo" src="https://user-images.githubusercontent.com/7910250/45997311-021c6000-c054-11e8-8725-a2220ff3b7ca.png">
<img width="305" alt="git_ ___src_test-repo" src="https://user-images.githubusercontent.com/7910250/45997324-0d6f8b80-c054-11e8-9658-06804bdc65b8.png">

This solution still leaves room for the user to accidentally click the `Undo Discard` button. It's right next to the `Stage All` button and, depending on styling and theme, the boundary between the two buttons may not be super visible, making it easier for users to mistakenly hit the `Undo discard` button when trying to hit the `Stage All` button.

Hiding the undo-discard functionality behind a context menu on the header makes it harder to accidentally trigger this action. See images in section above.

### Benefits

This UX makes it harder to accidentally undo the last discard.

### Possible Drawbacks

The undo discard functionality might be less discoverable since it is hidden behind a menu. Users may be confused by the disappearance of a button from the UI. 

### Applicable Issues

Issue -- #1242
Supersedes PR #1252
Fixes #1260 

### Metrics



- [x] Instrument `undoLastDiscard` method
- [x] Instrument discard methods
- [x] Instrument UI entry points (buttons, commands, header menu)

### Tests



Manual testing steps:
* click on <kbd>...</kbd> menu
* verify that undo last discard is greyed out when there is no discard history
* verify that "Discard Changes in Selected File" is pluralized correctly
* verify functionality of all menu items (discard all, discard selected files, undo last discard)
* verify that Stage/Unstage All buttons remain in UI and are disabled when there are no changes
* verify that header context menus no longer show "Stage All" and "Discard All" 

No context menu unit tests added as there is currently no way to write UI tests for context menus in electron --
https://github.com/electron/spectron/issues/21

Added metrics tests for recording the following:
* discarding all changes from the StagingView
* discarding selected changes from the StagingView
* undoing the last discard from the StagingView
* discarding selected lines from the FilePatchController
* undoing the last discard from the FilePatchController

### Documentation



- [ ] Update flight manual - https://github.com/atom/flight-manual.atom.io/pull/488 (add issue to merge this in project board for correct release)

### User Experience Research (Optional)



Some light UXR before shipping would be helpful. Especially to assess discoverability.
After shipping, we can monitor usage of the discard/undo discard functionality. It would be interesting to tweet about it and see if that affects feature usage. 

Questions:
- [x] How's the wording of the context menu items?
- [x] When there are no unstaged changes the `Stage All` button disappears and the ellipses icon moves to the right. This might look jarring. Maybe we should keep it fixed in the middle of the header, or place it to the right of the `Stage All` button. If we place it to the right then the `Stage All` button won't line up with the `Unstage All` button in the section below...
- [x] Would vertical ellipses look better than horizontal ellipses? Note that the merge conflict section uses the horizontal ellipses for its bulk resolve menu. Also, seems like the vertical ellipses icon is not available in the octicon set bundled with Atom.

/cc @simurai for 💭s on ☝️ 

TODO:
- [x] Conditionally show `Undo Last Discard` only if there is undo history
- [x] Show items in menus at all times, but make unclickable if no undo history 
- [x] Remove context menu items for header and put in kebab

Future steps:
- [ ] Consider prompting users to confirm undo discard if head has moved since last discard 
- [ ] Update flight manual when changes hit stable
